### PR TITLE
Allow multiple sessions/passphrases

### PR DESCRIPTION
**Pull request by tsusanka**
Updates #839. I'll add a few comments where I have some questions.

Notable changes:

- We send `session_id` only in response to Initialize. In response to GetFeatures we omit it. The rational is that we do not need a way to find out the active session id, all we need is to either receive a new one or confirm the current one is the right one in Initialize. However, I am not sure about this anymore, maybe it is not worth it?
- I have limited the sessions count to 10 for now as @prusnak mentioned. Seems reasonable or maybe even too high? But as far as we have enough space then why not.
- There is no way to forget all sessions in Trezor from the Host, since the ClearSession message is only locking device (as discussed). Are we okay with that?

### CODE REVIEW

File: legacy/firmware/config.c [hunk 1]
```diff
@@ -437,6 +447,7 @@ static void config_compute_u2froot(const char *mnemonic,
   memcpy(u2froot->private_key.bytes, node.private_key,
          sizeof(node.private_key));
   memzero(&node, sizeof(node));
+  memzero(&seed, sizeof(node));
```
Comment (tsusanka) [inline legacy/firmware/config.c]: There should be of course `sizeof(seed)`, I'll fix that.

File: legacy/firmware/config.c [hunk 2]
```diff
@@ -406,27 +407,36 @@ void config_init(void) {
...
 }
 
 static void config_compute_u2froot(const char *mnemonic,
                                    StorageHDNode *u2froot) {
   static CONFIDENTIAL HDNode node;
+  static CONFIDENTIAL uint8_t seed[64];
   char oldTiny = usbTiny(1);
-  mnemonic_to_seed(mnemonic, "", sessionSeed,
-                   get_u2froot_callback);  // BIP-0039
+  mnemonic_to_seed(mnemonic, "", seed, get_u2froot_callback);  // BIP-0039
```
Comment (tsusanka) [inline legacy/firmware/config.c]: @andrewkozlik I was a bit confused here. You have used sessionSeed here and called `session_clear` at the end, effectively destroying the current session. It seems alright, since it seems to be called only in load/reset/recovery/upgrade scanarios. Was it intentional? Why not simply declare a new array as I did?

File: legacy/firmware/config.c [hunk 3]
```diff
@@ -772,12 +784,69 @@ bool config_changeWipeCode(const char *pin, const char *wipe_code) {
...
+  uint8_t lowest_index = MAX_SESSIONS_COUNT;
+  uint32_t lowest = sessionLRUCounter;
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage < lowest) {
+      lowest = sessionsCache[i].usage;
+      lowest_index = i;
+    }
+  }
+  // TODO: abort if lowest_index == MAX_SESSIONS_COUNT?
+  // TODO: it should never happen, so it is more of "internal" check
```
Comment (tsusanka) [inline legacy/firmware/config.c]: Maybe not worth it?

File: legacy/firmware/config.h [hunk 4]
```diff
@@ -56,6 +56,13 @@ typedef struct {
   STORAGE_BYTES(public_key, 33);
 } StorageHDNode;
 
+typedef struct {
+  uint8_t id[32];
+  uint32_t usage;
+  uint8_t seed[64];
+  bool seedCached;  // TODO: should be secbool or ok?
```
Comment (tsusanka) [inline legacy/firmware/config.h]: `secbool` was undefined here, so I wasn't sure if I should use it.

File: core/src/apps/common/seed.py [hunk 5]
```diff
@@ -140,6 +140,8 @@ def derive_slip21_node_without_passphrase(path: list) -> Slip21Node:
         raise Exception("Device is not initialized")
     seed = cache.get(cache.APP_COMMON_SEED_WITHOUT_PASSPHRASE)
     if seed is None:
+        if not cache.is_session_started():
```
Comment (matejcik) [inline core/src/apps/common/seed.py]: This piece is interesting -- AFAICT, "node without passphrase" is not bound to a session, is that right?

I don't like that this piece of code starts its own session, and I also don't like that we cache and re-derive the same value up to 10 times.

My first instinct would be to cache it locally in seed.py, but that would mean that it doesn't get cleared on device wipe.
We could make the wipe call into seed.py to clear the cache, or we could special-case "node without passphrase" in cache (and do it the way it used to be done, with a special variable instead of a cache key).

The question is mostly whether there will ever be more data that is per-seed and not per-session.

File: core/src/storage/cache.py [hunk 6]
```diff
@@ -1,36 +1,68 @@
...
-def get_session_id() -> bytes:
-    global _cache_session_id
-    if not _cache_session_id:
-        _cache_session_id = random.bytes(32)
-    return _cache_session_id
+def _move_session_ids_queue(session_id: bytes) -> None:
+    # Move the LRU session ids queue.
+    if session_id in _session_ids:
+        _session_ids.remove(session_id)
+    elif len(_session_ids) == _MAX_SESSIONS_COUNT:
```
Comment (matejcik) [inline core/src/storage/cache.py]: i'd prefer a to call this `while len(_session_ids) >= _MAX_SESSIONS_COUNT:`

File: core/src/storage/cache.py [hunk 7]
```diff
@@ -1,36 +1,68 @@
...
+        remove_session_id = _session_ids.pop()
+        del _caches[remove_session_id]
+
+    _session_ids.insert(0, session_id)
+
+
+def start_session(received_session_id: bytes = None) -> bytes:
+    if received_session_id and received_session_id in _session_ids:
+        session_id = bytes(received_session_id)
```
Comment (matejcik) [inline core/src/storage/cache.py]: why the `bytes()`?

File: core/src/storage/cache.py [hunk 8]
```diff
@@ -1,36 +1,68 @@
...
 
 
 def get(key: int) -> Any:
-    return _cache.get(key)
+    if _active_session_id is None:
+        return None  # no session active
+    return _caches[_active_session_id].get(key)
 
 
 def clear() -> None:
```
Comment (matejcik) [inline core/src/storage/cache.py]: call this `clear_all` maybe?

File: core/tests/test_storage.cache.py [hunk 9]
```diff
@@ -44,38 +43,32 @@ def call_Initialize(**kwargs):
...
-        # supplying a different session ID clears the cache
-        self.assertNotEqual(new_session_id, features.session_id)
-        call_Initialize(session_id=new_session_id)
+        # supplying a different session ID starts a new cache
+        call_Initialize(session_id=b"A")
         self.assertIsNone(cache.get(KEY))
 
+        # but resuming a session loads the previous one
+        call_Initialize(session_id=session_id)
+        self.assertEqual(cache.get(KEY), "hello")
```
Comment (matejcik) [inline core/tests/test_storage.cache.py]: there should also be a test for two different values stored in two different sessions

File: core/src/apps/common/seed.py (hunk 5 above)
Comment (tsusanka) [reply core/src/apps/common/seed.py]: Yes this is something I wanted to discuss.
> AFAICT, "node without passphrase" is not bound to a session, is that right?

Good point. Maybe it indeed makes sense to have smth like `overall_cache` or `session_less_cache` (name TBD) in cache.py, which would store non-session-related items. In `cache.clear` we would clear it.

File: core/tests/test_storage.cache.py (hunk 9 above)
Comment (matejcik) [reply core/tests/test_storage.cache.py]: and a test that when sessions are exhausted, you can't get back the old item from the old session id. this is kinda-sorta subsumed in the above session id test, but i'd still like to have it separate

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 10]
```diff
@@ -54,24 +67,82 @@ def test_session_with_passphrase(client):
...
+    # creating one more will exceed the limit, the LRU item is at the moment the last one (see above)
+    # it should have been removed -> must ask for passphrase
+    _init_session(client)
+    _get_xpub(client, passphrase="XX")  # create one more
+    _init_session(client, session_id=session_ids[SESSIONS_STORED - 1])
+    _get_xpub(client, passphrase="whatever")  # the session is gone
+
+    # now the second to last is the next LRU
+    _init_session(client)
+    _get_xpub(client, passphrase="XXXX")
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py]: i'm not sure about how exactly to formulate the testcase, but something like ... "create a session that evicts, and then set the same passphrase as with the just-evicted, and check that the eviction did happen"

implementation-wise it is as simple as setting `passphrase="B"` here, but a comment explaining what i'm trying to achieve would help :)

File: tests/device_tests/test_session_id_and_passphrase.py [hunk 11]
```diff
@@ -95,24 +166,24 @@ def test_session_enable_passphrase(client):
     # We clear the session id now, so the passphrase should be asked.
     new_session_id = _init_session(client)
     assert session_id != new_session_id
-    assert _get_xpub(client, passphrase="A") == XPUB_PASSPHRASE_A
+    assert _get_xpub(client, passphrase="A") == XPUB_PASSPHRASES["A"]
 
 
 @pytest.mark.skip_ui
 @pytest.mark.setup_client(passphrase=True)
 def test_clear_session_passphrase(client):
```
Comment (matejcik) [inline tests/device_tests/test_session_id_and_passphrase.py]: judging by the name ISTM this testcase should have been removed before? now it is duplicating `test_session_with_passphrase`

File: core/src/storage/cache.py (hunk 6 above)
Comment (tsusanka) [reply core/src/storage/cache.py]: Mm you mean to protect against a case where we `_MAX_SESSIONS_COUNT + 2` items in `session_ids` for example?

Comment (matejcik) [reply core/src/storage/cache.py]: `+ 1` is enough for it to be relevant but yes, basically. there seems to be no downside to write the code more defensively, even though this should never happen

File: core/src/storage/cache.py (hunk 7 above)
Comment (tsusanka) [reply core/src/storage/cache.py]: Because it is used as a key to to the dict. And dict's key needs to be immutable, which `bytearray` is not. In other words, `__hash__` method is missing. Am I mistaken?

File: core/src/storage/cache.py (hunk 6 above)
Comment (tsusanka) [reply core/src/storage/cache.py]: Yes that was an example :). Ack, coming soon

File: core/src/storage/cache.py (hunk 7 above)
Comment (tsusanka) [reply core/src/storage/cache.py]: But if you are asking why am I receiving `bytearray` instead of `bytes` in the Initialize message, I'm not sure. Anyway I'll add it to the `handle_Initialize` function so the type hint is correct.

File: core/src/apps/common/seed.py (hunk 5 above)
Comment (tsusanka) [reply core/src/apps/common/seed.py]: Implemented in a4b5fdb6cea85d444d0d579fc9d20936eebf0791

File: core/src/storage/cache.py (hunk 7 above)
Comment (matejcik) [reply core/src/storage/cache.py]: yes, the second one :) converting this in handle_Initialize would be preferable, as the bytearrays are coming in from protobuf, but this code doesn't know that

File: core/src/storage/cache.py (hunk 6 above)
Comment (matejcik) [reply core/src/storage/cache.py]: i don't think the `else` is needed

File: tests/device_tests/test_session_id_and_passphrase.py (hunk 10 above)
Comment (tsusanka) [reply tests/device_tests/test_session_id_and_passphrase.py]: I am not sure I understand. We can of course add `passphrase="B"` here instead of `"XXXX"` but what does it test? It tests we prompt for passphrase "B" although it is just being removed. But since we do not send the sessions id and we cache based the sessons id, does it make any sense?

Comment (tsusanka) [reply tests/device_tests/test_session_id_and_passphrase.py]: Well anyway, it definitely _doesn't hurt_, so I'll add it :).

Comment (matejcik) [reply tests/device_tests/test_session_id_and_passphrase.py]: What it seems to me this test set obscures is the fact that _sessions are independent from passphrases_. Using `passphrase="B"` here hints at that, but you are right that it's not really _testing_ anything meaningful.

Perhaps a better choice is something like, create several sessions and not set a passphrase in them?
Or maybe something like this?
```
sessions = []
for _ in range(SESSIONS):
    sessions.append(_init_session(client))
    _get_xpub(client, passphrase="A")
for session_id in sessions:
	_init_session(client, session_id)
	_get_xpub(client, passphrase="A")
```
Do we have a different way to check that there's a different session cache, than using a passphrase?

File: legacy/firmware/config.c (hunk 2 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: This doesn't seem familiar at all and I can't see what you are talking about, because the force-push destroyed the references.

File: legacy/firmware/config.c [hunk 12]
```diff
@@ -120,11 +122,10 @@ be added to the storage u2f_counter to get the real counter value.
  * storage.u2f_counter + config_u2f_offset.
  * This corresponds to the number of cleared bits in the U2FAREA.
  */
-static secbool sessionSeedCached;
-static uint8_t CONFIDENTIAL sessionSeed[64];
+static CONFIDENTIAL Session sessionsCache[10];
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: Shouldn't the 10 be MAX_SESSIONS_COUNT?

File: legacy/firmware/config.c (hunk 2 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: I am basically talking about our current master, see https://github.com/trezor/trezor-firmware/blob/2e2f5fd421ea5db06604ea09c320bdc9028961bd/legacy/firmware/config.c#L426 .

File: legacy/firmware/config.c [hunk 13]
```diff
@@ -406,27 +407,36 @@ void config_init(void) {
...
 }
 
+void session_clearCache(Session *session) {
+  session->usage = 0;
+  memzero(session->id, sizeof(session->id));
+  memzero(session->seed, sizeof(session->seed));
+  session->seedCached = false;
+}
+
+void session_lock(void) { storage_lock(); }
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: I don't see the motivation behind this.

File: legacy/firmware/config.c (hunk 2 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]:
```
045ef22d (Saleem Rashid   2017-12-18 21:16:05 +0000 270) static void storage_compute_u2froot(const char* mnemonic, StorageHDNode *u2froot) {
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 271)        static CONFIDENTIAL HDNode node;
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 272)        char oldTiny = usbTiny(1);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 273)        mnemonic_to_seed(mnemonic, "", sessionSeed, get_u2froot_callback); // BIP-0039
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 274)        usbTiny(oldTiny);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 275)        hdnode_from_seed(sessionSeed, 64, NIST256P1_NAME, &node);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 276)        hdnode_private_ckd(&node, U2F_KEY_PATH);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 277)        u2froot->depth = node.depth;
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 278)        u2froot->child_num = U2F_KEY_PATH;
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 279)        u2froot->chain_code.size = sizeof(node.chain_code);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 280)        memcpy(u2froot->chain_code.bytes, node.chain_code, sizeof(node.chain_code));
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 281)        u2froot->has_private_key = true;
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 282)        u2froot->private_key.size = sizeof(node.private_key);
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 283)        memcpy(u2froot->private_key.bytes, node.private_key, sizeof(node.private_key));
bd660655 (Pavol Rusnak    2018-01-18 15:21:48 +0100 284)        memzero(&node, sizeof(node));
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 285)        session_clear(false); // invalidate seed cache
73edc7cb (Jochen Hoenicke 2017-12-13 10:53:27 +0100 286) }
```

File: legacy/firmware/config.c [hunk 14]
```diff
@@ -575,13 +587,13 @@ const uint8_t *config_getSeed(void) {
...
     char oldTiny = usbTiny(1);
-    mnemonic_to_seed(mnemonic, passphrase, sessionSeed,
+    mnemonic_to_seed(mnemonic, passphrase, activeSessionCache->seed,
                      get_root_node_callback);  // BIP-0039
     memzero(mnemonic, sizeof(mnemonic));
     memzero(passphrase, sizeof(passphrase));
     usbTiny(oldTiny);
-    sessionSeedCached = sectrue;
-    return sessionSeed;
+    activeSessionCache->seedCached = true;
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: Use sectrue.

**Comment by tsusanka** [comment]
Note: I believe there is a SIGSEGV when doing a reset. I will look into it.

File: legacy/firmware/config.c [hunk 15]
```diff
@@ -772,12 +784,69 @@ bool config_changeWipeCode(const char *pin, const char *wipe_code) {
...
+    }
+  }
+  // TODO: abort if lowest_index == MAX_SESSIONS_COUNT?
+  // TODO: it should never happen, so it is more of "internal" check
+  return lowest_index;
+}
+
+uint8_t session_findSession(const uint8_t *sessionId) {
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage != 0) {
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: Where do you initialize `usage` to 0? I guess you are calling `session_clear()` somewhere during startup, but I just can't find it anywhere.

File: legacy/firmware/config.c [hunk 16]
```diff
@@ -772,12 +784,69 @@ bool config_changeWipeCode(const char *pin, const char *wipe_code) {
...
-    random_buffer(sessionId, 32);
+uint8_t session_findLRU(void) {
+  uint8_t lowest_index = MAX_SESSIONS_COUNT;
+  uint32_t lowest = sessionLRUCounter;
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage < lowest) {
+      lowest = sessionsCache[i].usage;
+      lowest_index = i;
+    }
+  }
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: I find this really hard to read. Rename:
lowest -> oldest_use
lowest_index -> oldest_index
.usage -> .last_use

File: legacy/firmware/config.c [hunk 17]
```diff
@@ -772,12 +784,69 @@ bool config_changeWipeCode(const char *pin, const char *wipe_code) {
...
 }
 
-const uint8_t *session_getSessionId(void) {
-  if (!sessionIdCached) {
-    random_buffer(sessionId, 32);
+uint8_t session_findLRU(void) {
+  uint8_t lowest_index = MAX_SESSIONS_COUNT;
+  uint32_t lowest = sessionLRUCounter;
+  for (uint8_t i = 0; i < MAX_SESSIONS_COUNT; i++) {
+    if (sessionsCache[i].usage < lowest) {
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: Better use `<=`.

File: legacy/firmware/config.c (hunk 3 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: It will happen if `sessionLRUCounter` overflows.

File: legacy/firmware/config.c [hunk 18]
```diff
@@ -772,12 +784,69 @@ bool config_changeWipeCode(const char *pin, const char *wipe_code) {
...
-  return sessionId;
+  if (session_index != MAX_SESSIONS_COUNT) {
+    // session found in currently stored sessions
+    sessionLRUCounter++;
+    sessionsCache[session_index].usage = sessionLRUCounter;
+    activeSessionCache = sessionsCache + session_index;
+    return activeSessionCache->id;
+  } else {
+    // if not found or session_id not provided try to find an empty one
+    session_index = session_findEmpty();
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: Do we need `session_findEmpty`? If you call `session_findLRU` and there is an empty session, then it will return the empty session won't it?

File: legacy/firmware/config.c [hunk 19]
```diff
@@ -772,12 +784,69 @@ bool config_changeWipeCode(const char *pin, const char *wipe_code) {
...
+
+  // now it is guaranteed `session_index` is the place to start a new session
+  sessionLRUCounter++;
+  session_clearCache(sessionsCache + session_index);
+
+  sessionsCache[session_index].usage = sessionLRUCounter;
+  random_buffer(sessionsCache[session_index].id, 32);
+
+  activeSessionCache = sessionsCache + session_index;
+  return activeSessionCache->id;
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: I find this a bit hard to follow. I think it can be written more cleanly:
```
uint8_t *session_startSession(const uint8_t *received_session_id) {
  int session_index = MAX_SESSIONS_COUNT;

  if (received_session_id != NULL) {
    session_index = session_findSession(received_session_id);
  }

  if (session_index == MAX_SESSIONS_COUNT) {
    // Session not found in cache. Use the least recently used session in the cache.
    session_index = session_findLRU();
    session_clearCache(sessionsCache + session_index);
    random_buffer(sessionsCache[session_index].id, 32);
  }

  sessionLRUCounter++;
  sessionsCache[session_index].usage = sessionLRUCounter;
  activeSessionCache = sessionsCache + session_index;
  return activeSessionCache->id;
}
```

File: legacy/firmware/config.h (hunk 4 above)
Comment (andrewkozlik) [reply legacy/firmware/config.h]: Why couldn't we include secbool? Actually, why does Session need to be declared in the header? Isn't it just an internal structure?

File: legacy/firmware/config.h [hunk 20]
```diff
@@ -108,7 +117,10 @@ bool config_getPassphraseProtection(bool *passphrase_protection);
 bool config_getHomescreen(uint8_t *dest, uint16_t dest_size);
 void config_setHomescreen(const uint8_t *data, uint32_t size);
 
-const uint8_t *session_getSessionId(void);
+uint8_t session_findLRU(void);
+uint8_t session_findSession(const uint8_t *sessionId);
+uint8_t session_findEmpty(void);
```
Comment (andrewkozlik) [inline legacy/firmware/config.h]: Shouldn't these functions be for internal use only? Or is there some bigger picture I am not seeing?

File: legacy/firmware/config.c [hunk 21]
```diff
@@ -575,13 +587,13 @@ const uint8_t *config_getSeed(void) {
       }
     }
     char oldTiny = usbTiny(1);
-    mnemonic_to_seed(mnemonic, passphrase, sessionSeed,
+    mnemonic_to_seed(mnemonic, passphrase, activeSessionCache->seed,
```
Comment (andrewkozlik) [inline legacy/firmware/config.c]: Can we be absolutely sure that `activeSessionCache` is not` NULL`?

File: legacy/firmware/config.c (hunk 16 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: I agree with the first two. About the third one I am not sure. How about `usage_counter`? Or `use_count`? `last_use` feels like the highest number is the last used, but it is the opposite.

File: legacy/firmware/config.c (hunk 18 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: Good catch. That should work as far as there is `<=` in `session_findLRU` as you also suggested. I will also add:

```
if (sessionsCache[i].usage == 0) {
  return i;
}
```

as a small optimalization.

Comment (tsusanka) [reply legacy/firmware/config.c]: I am wondering, should we rename `session_findLRU` to reflect the fact it finds an empty session as well? It is a LRU item as well after all, so maybe not worth it?

Suggestions:
- session_findIndex
- session_findPosition
- session_nextCacheIndex

File: legacy/firmware/config.c (hunk 19 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: Yes, that looks better.

File: legacy/firmware/config.h (hunk 20 above)
Comment (tsusanka) [reply legacy/firmware/config.h]: Indeed, they should. I have a bad habit of putting everything to config.h automatically. Will move to config.c.

File: legacy/firmware/config.c (hunk 21 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: Turns out we can't, it is the reason the segfault occurred that I've mentioned. If the Host sends Initialize before any call this should not happen, but we of course can't depend on that.

File: legacy/firmware/config.c (hunk 15 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: That is a very good point. I think I did not. I have added it to config.c to a `config_init` function, which is called from the main `trezor.c`.

File: legacy/firmware/config.h (hunk 4 above)
Comment (tsusanka) [reply legacy/firmware/config.h]: Moved to config.c.

File: legacy/firmware/config.c (hunk 3 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: Indeed. Since `sessionLRUCounter` is 32 bytes, it seems like a fair assumption this should not happen?

File: legacy/firmware/config.c (hunk 13 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: It's basically just a syntax sugar. `session_lock` is called in places where you work with sessions, however not with storage directly. Also, it seemed reasonable to provide a function that locks the session, even though it is currently implemented as locking the storage only. It can be extended easily in future.

**Comment by tsusanka** [comment]
Review from @andrewkozlik implemented in a763883f or commented.

File: legacy/firmware/config.c (hunk 2 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: Ahh, I did not dig deep enough. @prusnak do you remember anything about this?

File: legacy/firmware/config.c (hunk 3 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: Oh, I am not worried about regular use, but attacker could use this. I don't know how many messages per second we can process on the USB, so I don't know how realistic that is.

Comment (andrewkozlik) [reply legacy/firmware/config.c]: `ensure(lowest_index < MAX_SESSIONS_COUNT, NULL)`

File: legacy/firmware/config.c (hunk 13 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: 1. I don't like wrapper functions that don't add any functionality, because they just obfuscate the code. Of course, if we know we will be adding some functionality here in the future, then that's perfectly fine. I just couldn't to think of anything that would make sense here.
2. The name doesn't make much sense to me. It doesn't lock a session. It locks the device, i.e. all sessions and possibly more.

File: legacy/firmware/config.c (hunk 18 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: Well I don't like the name of the function either, but mainly because of the abbreviation. I would have called it `session_find_least_recent()`, but I can accept the abbreviation, since it's only a local helper function and the comment explains what it does.

In any case I think the function name clearly includes finding an unused session. Any session which has not been used previously can be considered least recent.

File: legacy/firmware/config.c (hunk 13 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: I mean, I feel this does not have anything to do with sessions and I just find it misleading.

... I see this is related to the naming of the protobuf messages and https://github.com/trezor/trezor-firmware/issues/819. So if we really need this function (which I doubt), then let's call it `lock_device()` or something.

File: legacy/firmware/config.c (hunk 16 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: I am not sure I understand. I thought `.usage` is meant to record when the session was last used. It doesn't count the number of times it was used. If it did, then you would be discarding the least used session, and we agreed we don't want that.

BTW, I am open to any equivalent alternatives for the first two, e.g. `least_recent_use` and `least_recent_index`, as long as the name says something about the variable. I can't think of any alternatives for the third though.

File: legacy/firmware/config.c (hunk 3 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: Great. Since `ensure` uses `secbool` I have used `ensure(sectrue * (oldest_index < MAX_SESSIONS_COUNT), NULL);`

Comment (tsusanka) [reply legacy/firmware/config.c]: Ok, great. `ensure` as defined in `common.h` uses `secbool`, so I will use `ensure(sectrue * (oldest_index < MAX_SESSIONS_COUNT), NULL);` correct<

File: legacy/firmware/config.c (hunk 18 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: I agree. Let's go for `session_find_least_recent`. Actually, to stick to the notation let's name it `session_findLeastRecent`.

File: legacy/firmware/config.c (hunk 16 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: Ok, I believe those are the best:

lowest -> least_recent_use
lowest_index -> least_recent_index
.usage -> .last_use

Comment (tsusanka) [reply legacy/firmware/config.c]: It would be nice to rename `sessionLRUCounter` as well to fit this new names. `sessionUseCounter`?

File: legacy/firmware/config.c (hunk 13 above)
Comment (tsusanka) [reply legacy/firmware/config.c]: 1. I agree that wrapper functions are not usually needed. But I believe in this case there is a real benefit. By providing a function in `config.c` to lock the device, we provide a clear interface to perform such action. The fact that it is currently only `storage_lock` is not important. A new developer/3rd party will not realize `storage_lock == lock_device`.
2. I agree the name is not good and I agree with `lock_device`.

Comment (tsusanka) [reply legacy/firmware/config.c]: Well, actually this should be `config_lockDevice` to fit the others.

File: legacy/firmware/config.c (hunk 3 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: Yes, you are right.

File: legacy/firmware/config.c (hunk 13 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: `config_lockDevice()` is perfect.

File: legacy/firmware/config.c (hunk 16 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: I totally agree.

File: legacy/firmware/config.c (hunk 18 above)
Comment (andrewkozlik) [reply legacy/firmware/config.c]: Great!

**Comment by tsusanka** [comment]
Second fixes in 33bfcb053d501389de094a011e0ad4f6f67150b4 . @andrewkozlik please check it. I will squash and rebase on top of master then.

File: core/src/storage/cache.py [hunk 22]
```diff
@@ -1,36 +1,83 @@
...
-def get_session_id() -> bytes:
-    global _cache_session_id
-    if not _cache_session_id:
-        _cache_session_id = random.bytes(32)
-    return _cache_session_id
+def _move_session_ids_queue(session_id: bytes) -> None:
+    # Move the LRU session ids queue.
+    if session_id in _session_ids:
+        _session_ids.remove(session_id)
+    else:
```
Comment (matejcik) [inline core/src/storage/cache.py]: this `else` should not be here IMO

Comment (tsusanka) [reply core/src/storage/cache.py]: You mean because the `while` loop is not evaluated in that case? It still seems more readable if present imho

Comment (matejcik) [reply core/src/storage/cache.py]: the `while` loop is defending for a case where `len(_session_ids) > _MAX_SESSIONS_COUNT`
that can be true even in case we removed an existing session id

Comment (tsusanka) [reply core/src/storage/cache.py]: Hm I guess if we have the `while` there we indeed should do it as you suggest. Will fix.

File: core/src/storage/cache.py [hunk 23]
```diff
@@ -1,36 +1,82 @@
...
+APP_MONERO_LIVE_REFRESH = 2
+
+# Keys that are valid across sessions
+APP_COMMON_SEED_WITHOUT_PASSPHRASE = 1 | _SESSIONLESS_FLAG
+
 
-_cache_session_id = None  # type: Optional[bytes]
-_cache = {}
+_active_session_id = None  # type: Optional[bytes]
+_caches = {}  # type: Dict[bytes, Dict[int, Any]]
```
Comment (matejcik) [inline core/src/storage/cache.py]: nit: perhaps we should have `CacheDict = Dict[int, Any]`, and use that as the type hint
